File: backend/app/engine/layers/triggers.py
"""
Layer 4: Trigger Detector.

Checks for actionable entry triggers using multiple confirmation signals.
Requires a configurable number of trigger types to fire within a lookback
window for a confirmed entry.

Trigger types:
  A. MACD crossover in trend direction
  B. RSI crossing midline (50)
  C. Stochastic leaving overbought/oversold
  D. Engulfing candlestick pattern
  E. Price reclaiming zone
"""

from dataclasses import dataclass, field

import pandas as pd

from app.engine.indicators import compute_macd, compute_rsi, compute_stochastic
from app.engine.layers.trend import Trend, TrendResult
from app.engine.layers.zones import EntryZone


@dataclass
class TriggerResult:
    confirmed: bool
    price: float
    confirmations: list[str] = field(default_factory=list)


class TriggerDetector:
    """
    Layer 4: Checks for actionable entry triggers.

    Evaluates 5 independent trigger conditions over a configurable lookback
    window and requires a configurable number of confirmations.
    """

    def __init__(
        self,
        min_confirmations: int = 2,
        lookback: int = 1,
    ):
        self.min_confirmations = min_confirmations
        self.lookback = lookback

    def check(
        self,
        candles: pd.DataFrame,
        zone: EntryZone,
        trend: TrendResult,
    ) -> TriggerResult:
        """Evaluate all trigger conditions against candle data."""
        confirmations: list[str] = []
        price = float(candles["close"].iloc[-1])

        if self._check_macd_crossover(candles, trend, self.lookback):
            confirmations.append("macd_crossover")

        if self._check_rsi_midline(candles, trend, self.lookback):
            confirmations.append("rsi_midline_cross")

        if self._check_stochastic_exit(candles, trend, self.lookback):
            confirmations.append("stochastic_exit_extreme")

        if self._check_engulfing(candles, trend, self.lookback):
            confirmations.append("engulfing_candle")

        if self._check_zone_reclaim(candles, zone, trend, self.lookback):
            confirmations.append("zone_reclaim")

        confirmed = len(confirmations) >= self.min_confirmations

        return TriggerResult(
            confirmed=confirmed,
            price=price,
            confirmations=confirmations,
        )

    # ------------------------------------------------------------------
    # Trigger A: MACD crossover in trend direction
    # ------------------------------------------------------------------
    @staticmethod
    def _check_macd_crossover(
        candles: pd.DataFrame, trend: TrendResult, lookback: int = 1,
    ) -> bool:
        """MACD line crosses signal line in the direction of the trend."""
        if len(candles) < 35:
            return False

        macd_line, signal_line, _ = compute_macd(candles["close"])

        for offset in range(lookback):
            idx = -1 - offset
            prev_idx = idx - 1
            if abs(prev_idx) > len(macd_line):
                break

            cur_macd = macd_line.iloc[idx]
            prev_macd = macd_line.iloc[prev_idx]
            cur_signal = signal_line.iloc[idx]
            prev_signal = signal_line.iloc[prev_idx]

            nans = (
                pd.isna(cur_macd) or pd.isna(prev_macd)
                or pd.isna(cur_signal) or pd.isna(prev_signal)
            )
            if nans:
                continue

            if trend.direction == Trend.BULLISH:
                if prev_macd <= prev_signal and cur_macd > cur_signal:
                    return True
            elif trend.direction == Trend.BEARISH:
                if prev_macd >= prev_signal and cur_macd < cur_signal:
                    return True

        return False

    # ------------------------------------------------------------------
    # Trigger B: RSI crossing midline (50)
    # ------------------------------------------------------------------
    @staticmethod
    def _check_rsi_midline(
        candles: pd.DataFrame, trend: TrendResult, lookback: int = 1,
    ) -> bool:
        """RSI crosses the 50 midline in the direction of the trend."""
        if len(candles) < 20:
            return False

        rsi = compute_rsi(candles["close"], period=14)

        for offset in range(lookback):
            idx = -1 - offset
            prev_idx = idx - 1
            if abs(prev_idx) > len(rsi):
                break

            cur_rsi = rsi.iloc[idx]
            prev_rsi = rsi.iloc[prev_idx]

            if pd.isna(cur_rsi) or pd.isna(prev_rsi):
                continue

            if trend.direction == Trend.BULLISH:
                if prev_rsi <= 50 and cur_rsi > 50:
                    return True
            elif trend.direction == Trend.BEARISH:
                if prev_rsi >= 50 and cur_rsi < 50:
                    return True

        return False

    # ------------------------------------------------------------------
    # Trigger C: Stochastic leaving overbought/oversold
    # ------------------------------------------------------------------
    @staticmethod
    def _check_stochastic_exit(
        candles: pd.DataFrame, trend: TrendResult, lookback: int = 1,
    ) -> bool:
        """Stochastic %K exits overbought (>80) or oversold (<20) territory."""
        if len(candles) < 20:
            return False

        slowk, _ = compute_stochastic(candles)

        for offset in range(lookback):
            idx = -1 - offset
            prev_idx = idx - 1
            if abs(prev_idx) > len(slowk):
                break

            cur_k = slowk.iloc[idx]
            prev_k = slowk.iloc[prev_idx]

            if pd.isna(cur_k) or pd.isna(prev_k):
                continue

            if trend.direction == Trend.BULLISH:
                if prev_k <= 20 and cur_k > 20:
                    return True
            elif trend.direction == Trend.BEARISH:
                if prev_k >= 80 and cur_k < 80:
                    return True

        return False

    # ------------------------------------------------------------------
    # Trigger D: Engulfing candlestick pattern
    # ------------------------------------------------------------------
    @staticmethod
    def _check_engulfing(
        candles: pd.DataFrame, trend: TrendResult, lookback: int = 1,
    ) -> bool:
        """Bullish or bearish engulfing pattern within lookback window."""
        if len(candles) < 2:
            return False

        for offset in range(lookback):
            idx = -1 - offset
            prev_idx = idx - 1
            if abs(prev_idx) >= len(candles):
                break

            prev_open = candles["open"].iloc[prev_idx]
            prev_close = candles["close"].iloc[prev_idx]
            cur_open = candles["open"].iloc[idx]
            cur_close = candles["close"].iloc[idx]

            if trend.direction == Trend.BULLISH:
                prev_bearish = prev_close < prev_open
                cur_bullish = cur_close > cur_open
                engulfs = cur_open <= prev_close and cur_close >= prev_open
                if prev_bearish and cur_bullish and engulfs:
                    return True
            elif trend.direction == Trend.BEARISH:
                prev_bullish = prev_close > prev_open
                cur_bearish = cur_close < cur_open
                engulfs = cur_open >= prev_close and cur_close <= prev_open
                if prev_bullish and cur_bearish and engulfs:
                    return True

        return False

    # ------------------------------------------------------------------
    # Trigger E: Price reclaiming zone
    # ------------------------------------------------------------------
    @staticmethod
    def _check_zone_reclaim(
        candles: pd.DataFrame, zone: EntryZone, trend: TrendResult,
        lookback: int = 1,
    ) -> bool:
        """Price moves back into the entry zone from outside it."""
        if len(candles) < 2:
            return False

        for offset in range(lookback):
            idx = -1 - offset
            prev_idx = idx - 1
            if abs(prev_idx) >= len(candles):
                break

            prev_close = candles["close"].iloc[prev_idx]
            cur_close = candles["close"].iloc[idx]

            if trend.direction == Trend.BULLISH:
                if prev_close < zone.lower and cur_close >= zone.lower:
                    return True
            elif trend.direction == Trend.BEARISH:
                if prev_close > zone.upper and cur_close <= zone.upper:
                    return True

        return False

File: supabase/functions/_shared/engine/triggers.ts
/**
 * Layer 4: Trigger Detector.
 *
 * Checks for actionable entry triggers using multiple confirmation signals.
 * Requires a configurable number of trigger types to fire within a lookback
 * window for a confirmed entry.
 *
 * Trigger types:
 *   A. MACD crossover in trend direction
 *   B. RSI crossing midline (50)
 *   C. Stochastic leaving overbought/oversold
 *   D. Engulfing candlestick pattern
 *   E. Price reclaiming zone
 */

import {
  calcMACD, calcRSI, calcStochastic, extractArrays, type OHLCV,
} from '../indicators.ts'
import { Trend, type TrendResult } from './trend.ts'
import type { EntryZone } from './zones.ts'

// ── Trigger result ─────────────────────────────────────────────
export interface TriggerResult {
  confirmed: boolean
  price: number
  confirmations: string[]
}

// ── Trigger Detector ───────────────────────────────────────────
export class TriggerDetector {
  private minConfirmations: number
  private lookback: number

  constructor(minConfirmations = 2, lookback = 1) {
    this.minConfirmations = minConfirmations
    this.lookback = lookback
  }

  check(candles: OHLCV[], zone: EntryZone, trend: TrendResult): TriggerResult {
    const confirmations: string[] = []
    const price = candles[candles.length - 1].close

    if (this._checkMacdCrossover(candles, trend, this.lookback)) {
      confirmations.push('macd_crossover')
    }
    if (this._checkRsiMidline(candles, trend, this.lookback)) {
      confirmations.push('rsi_midline_cross')
    }
    if (this._checkStochasticExit(candles, trend, this.lookback)) {
      confirmations.push('stochastic_exit_extreme')
    }
    if (this._checkEngulfing(candles, trend, this.lookback)) {
      confirmations.push('engulfing_candle')
    }
    if (this._checkZoneReclaim(candles, zone, trend, this.lookback)) {
      confirmations.push('zone_reclaim')
    }

    return {
      confirmed: confirmations.length >= this.minConfirmations,
      price,
      confirmations,
    }
  }

  // ── Trigger A: MACD crossover ────────────────────────────────
  private _checkMacdCrossover(
    candles: OHLCV[], trend: TrendResult, lookback: number,
  ): boolean {
    if (candles.length < 35) return false

    const { closes } = extractArrays(candles)
    const macd = calcMACD(closes)

    for (let offset = 0; offset < lookback; offset++) {
      const idx = macd.length - 1 - offset
      const prevIdx = idx - 1
      if (prevIdx < 0) break

      const curMacd = macd[idx].MACD
      const prevMacd = macd[prevIdx].MACD
      const curSignal = macd[idx].signal
      const prevSignal = macd[prevIdx].signal

      if (
        curMacd === undefined || prevMacd === undefined ||
        curSignal === undefined || prevSignal === undefined
      ) continue

      if (trend.direction === Trend.BULLISH) {
        if (prevMacd <= prevSignal && curMacd > curSignal) return true
      } else if (trend.direction === Trend.BEARISH) {
        if (prevMacd >= prevSignal && curMacd < curSignal) return true
      }
    }
    return false
  }

  // ── Trigger B: RSI crossing midline (50) ─────────────────────
  private _checkRsiMidline(
    candles: OHLCV[], trend: TrendResult, lookback: number,
  ): boolean {
    if (candles.length < 20) return false

    const { closes } = extractArrays(candles)
    const rsi = calcRSI(closes, 14)

    for (let offset = 0; offset < lookback; offset++) {
      const idx = rsi.length - 1 - offset
      const prevIdx = idx - 1
      if (prevIdx < 0) break

      const curRsi = rsi[idx]
      const prevRsi = rsi[prevIdx]

      if (curRsi === undefined || prevRsi === undefined) continue
      if (isNaN(curRsi) || isNaN(prevRsi)) continue

      if (trend.direction === Trend.BULLISH) {
        if (prevRsi <= 50 && curRsi > 50) return true
      } else if (trend.direction === Trend.BEARISH) {
        if (prevRsi >= 50 && curRsi < 50) return true
      }
    }
    return false
  }

  // ── Trigger C: Stochastic leaving overbought/oversold ────────
  private _checkStochasticExit(
    candles: OHLCV[], trend: TrendResult, lookback: number,
  ): boolean {
    if (candles.length < 20) return false

    const { highs, lows, closes } = extractArrays(candles)
    const stoch = calcStochastic(highs, lows, closes)

    for (let offset = 0; offset < lookback; offset++) {
      const idx = stoch.length - 1 - offset
      const prevIdx = idx - 1
      if (prevIdx < 0) break

      const curK = stoch[idx].k
      const prevK = stoch[prevIdx].k

      if (curK === undefined || prevK === undefined) continue

      if (trend.direction === Trend.BULLISH) {
        if (prevK <= 20 && curK > 20) return true
      } else if (trend.direction === Trend.BEARISH) {
        if (prevK >= 80 && curK < 80) return true
      }
    }
    return false
  }

  // ── Trigger D: Engulfing candlestick pattern ─────────────────
  private _checkEngulfing(
    candles: OHLCV[], trend: TrendResult, lookback: number,
  ): boolean {
    if (candles.length < 2) return false

    for (let offset = 0; offset < lookback; offset++) {
      const idx = candles.length - 1 - offset
      const prevIdx = idx - 1
      if (prevIdx < 0) break

      const prev = candles[prevIdx]
      const cur = candles[idx]

      if (trend.direction === Trend.BULLISH) {
        const prevBearish = prev.close < prev.open
        const curBullish = cur.close > cur.open
        const engulfs = cur.open <= prev.close && cur.close >= prev.open
        if (prevBearish && curBullish && engulfs) return true
      } else if (trend.direction === Trend.BEARISH) {
        const prevBullish = prev.close > prev.open
        const curBearish = cur.close < cur.open
        const engulfs = cur.open >= prev.close && cur.close <= prev.open
        if (prevBullish && curBearish && engulfs) return true
      }
    }
    return false
  }

  // ── Trigger E: Price reclaiming zone ─────────────────────────
  private _checkZoneReclaim(
    candles: OHLCV[], zone: EntryZone, trend: TrendResult, lookback: number,
  ): boolean {
    if (candles.length < 2) return false

    for (let offset = 0; offset < lookback; offset++) {
      const idx = candles.length - 1 - offset
      const prevIdx = idx - 1
      if (prevIdx < 0) break

      const prevClose = candles[prevIdx].close
      const curClose = candles[idx].close

      if (trend.direction === Trend.BULLISH) {
        if (prevClose < zone.lower && curClose >= zone.lower) return true
      } else if (trend.direction === Trend.BEARISH) {
        if (prevClose > zone.upper && curClose <= zone.upper) return true
      }
    }
    return false
  }
}
